core/src/main/java/io/questdb/griffin/SqlOptimiser.java

```diff
@@ -3801,6 +3801,126 @@ private ExpressionNode rewriteAggregate(ExpressionNode agg, QueryModel model) {
+            ObjList<QueryColumn> queryColumns = parent.getBottomUpColumns();
+            QueryModel nestedModel = parent.getNestedModel();
+            CharSequence designatedTimestampColumn = null;
+
+            /**if FIRST/LAST/min/max(column) does not contain designated-timestamp column
+             * OR
+             * another column present in select clause
+             * then don't apply optimisations to base model
+             */
+            if (nestedModel.getTimestamp() == null || queryColumns.size() > 1) {
+                rewriteGroupByForFirstLastMaxMinAggregateFunctions(nestedModel);
+                return parent;
+            }
+
+            //get designated timestamp column name
+            designatedTimestampColumn = nestedModel.getTimestamp().token;
+            QueryColumn column = queryColumns.get(0);
+            ExpressionNode ast = column.getAst();
+            CharSequence token = null;
+            CharSequence rhs = null;
+            if (ast != null) {
+                token = ast.token;
+                rhs = ast.rhs == null ? null : ast.rhs.token;
+            }
+
+            /**
+             Type 1 Optimisations: Will be optimised by adding order by clause and changing model type
+             Type 2 Optimisations: Will be done only by changing model type to prevent invoking group by workers
+             */
+            int optimisationType = 0;
+            if (rhs != null && ast.type == 8 && Chars.equals(designatedTimestampColumn, rhs)) {
+                if (Chars.equalsIgnoreCase(token, "LAST") || Chars.equalsIgnoreCase(token, "MAX"))
+                    optimisationType = 1;
+                else if (Chars.equalsIgnoreCase(token, "FIRST") || Chars.equalsIgnoreCase(token, "MIN"))
+                    optimisationType = 2;
+            }
+
+            /**
+             Core logic for issue #4231 will be applicable under two conditions
+             Condition1: QueryColumn contains FIRST/LAST/max/min function keyword
+             Condition2: QueryColumn for LAST/max keyword should be same as designatedTimestamp column
+             Core logic for changing query plan
+             - Add limit 1 via expressionNode to parent model
+             - Change model to select-choose from select-group-by
+             - Column alias by default is LAST/max if alias is not provided, replace with original column name
+             - Add order by clause to nested model
+             Call this recursively for nested models
+             */
+            if (optimisationType == 1 || optimisationType == 2) {
+                QueryModel newNestedModel = queryModelPool.next();
+                ExpressionNode lowerLimitNode = expressionNodePool.next();
+                lowerLimitNode.token = "1";
+                lowerLimitNode.type = 2;
+                parent.setLimit(lowerLimitNode, null);
+
+                //change model type to select-choose
+                parent.setSelectModelType(1);
```

Comment:
Hardcoded constants aren't readable:
```suggestion
                parent.setSelectModelType(QueryModel.SELECT_MODEL_CHOOSE);
```